Changes in the handling of reveled message indexes (#14)
decentralized-identity/bbs-signature: spec.md

## SpkGen

A signature proof of knowledge generating algorithm that creates a zero-knowledge proof of knowledge of a signature while selectively disclosing messages from a signature given a vector of messages, a vector of indices of the revealed messages, the signer's public key, and a nonce.

```
spk = SpkGen(PK, (msg[1],...,msg[L]), RIdxs, signature, nonce)
```

Inputs:

- PK, octet string in output form from SkToPk, DpkToPk
- (msg\[1\],...,msg\[L\]), octet strings (messages in input to Sign).

> - RIdxs, integers vector (indices of revealed messages).

Reviewer: ```suggestion
- RIdxs, vector of unsigned integers (indices of revealed messages).
```

1. (A, e, s) = signature

2. (w, h0, h\[1\],...,h\[L\]) = PK

3. (i1, i2,..., iR) = RIdxs

4. if subgroup\_check(A) is INVALID abort

5. if KeyValidate(PK) is INVALID abort

6. b = commitment + h0 \* s + h\[1\] \* msg\[1\] + ... + h\[L\] \* msg\[L\]

7. r1 = H(PRF(8\*ceil(log2(r)))) mod r

8. r2 = H(PRF(8\*ceil(log2(r)))) mod r

9. e\~ = H(PRF(8\*ceil(log2(r)))) mod r

10. r2\~ = H(PRF(8\*ceil(log2(r)))) mod r

11. r3\~ = H(PRF(8\*ceil(log2(r)))) mod r

12. s\~ = H(PRF(8\*ceil(log2(r)))) mod r

13. r3 = r1 ^ -1 mod r

14. for i in RIdxs m\~\[i\] = H(PRF(8\*ceil(log2(r)))) mod r

15. A' = A \* r1

16. Abar = A' \* -e + b \* r1

17. d = b \* r1 + h0 \* -r2

18. s' = s - r2 \* r3

19. C1 = A' \* e\~ + h0 \* r2\~

20. C2 = d \* r3\~ + h0 \* s\~ + h\[i1\] \* m\~\[i1\] + ... + h\[iR\] \* m\~\[iR\]

21. c = H(Abar || A' || h0 || C1 || d || h0 || h\[i1\] || ... || h\[iR\] || C2 || nonce)

22. e^ = e\~ + c \* e

23. r2^ = r2\~ - c \* r2

24. r3^ = r3\~ + c \* r3

25. s^ = s\~ - c \* s'

26. for i in RIdxs: m^\[i\] = m\~\[i\] - c \* msg\[i\]

27. spk = ( A', Abar, d, C1, e^, r2^, C2, r3^, s^, (m^\[i1\], ..., m^\[iR\]) )

28. return spk

> How a signature is to be encoded is not covered by the BBS+ spec itshelf. A common way is is to use a DER-encoded ASN.1 SEQUENCE for the proof value and append to that the list of indexes of revealed messeges, creating the following structure,

Reviewer: ```suggestion
How a signature is to be encoded is not covered by this document. (TODO perhaps add some additional information in the appendix)
```

+-------------------------------------+--------------------------------------------+------------------------------------------+-------------------------+

| Total messages N (16bit)| Revealed messages N (16bit) | Revealed messages indexes | remaining proof |

+-------------------------------------+--------------------------------------------+------------------------------------------+-------------------------+

## SpkVerify

SpkVerify checks if a signature proof of knowledge is VALID given the proof, the signer's public key, a vector of revealed messages, a vector with the indices of these revealed messages, and the nonce used in SpkGen.

```
result = SpkVerify(spk, PK, (Rmsg[1],..., Rmsg[R]), RIdxs, nonce)
```

Inputs:

- spk, octet string.
- PK, octet string in output form from SkToPk, DpkToPk.
- (Rmsg\[1\], ..., Rmsg\[R\]), octet strings (revealed messages).

> - RIdxs, integers vector (indices of revealed messages).

Reviewer: ```suggestion
- RIdxs, vector of unsigned integers (indices of revealed messages).
```

1. if KeyValidate(PK) is INVALID

2. (i1, i2, ..., iR) = RIndxs

3. (A', Abar, d, C1, e^, r2^, C2, r3^, s^, (m^\[i1\],...,m^\[iR\])) = spk

4. (w, h0, h\[1\],...,h\[L\]) = PK

5. if A' == 1 return INVALID

6. X1 = e(A', w)

7. X2 = e(Abar, P2)

8. if X1 != X2 return INVALID

9. c = H(Abar || A' || h0 || C1 || d || h0 || h\[i1\] || ... || h\[iR\] || C2 || nonce)

10. T1 = Abar - d

11. T2 = P1 + h\[i1\] \* Rmsg\[1\] + ... + h\[iR\] \* Rmsg\[R\]

12. Y1 = A' \* e^ + h0 \* r2^ + T1 \* c

> 13. Y2 = d \* r3^ + h0 \* s^ + h\[ii\] \* m^\[i1\] + ... + h\[iR\] \* m^\[iR\] - T2 \* c

Reviewer: ```suggestion
13. Y2 = d \* r3^ + h0 \* s^ + h\[i1\] \* m^\[i1\] + ... + h\[iR\] \* m^\[iR\] - T2 \* c
```
